sdks/python/apache_beam/ml/transforms/handlers.py

```diff
@@ -0,0 +1,436 @@
+      beam.Row, Dict[str, np.ndarray], pa.RecordBatch]]:
+    """
+    This method also computes the required dataset metadata for the tft
+    AnalyzeDataset/TransformDataset step.
+
+    This method uses tensorflow_transform's Analyze step to produce the
+    artifacts and Transform step to apply the transforms on the data.
+    Artifacts are only produced if the artifact_mode is set to `produce`.
+    If artifact_mode is set to `consume`, then the artifacts are read from the
+    artifact_location, which was previously used to store the produced
+    artifacts.
+    """
+    if self.artifact_mode == ArtifactMode.PRODUCE:
+      # If we are computing artifacts, we should fail for windows other than
+      # default windowing since for example, for a fixed window, each window can
+      # be treated as a separate dataset and we might need to compute artifacts
+      # for each window. This is not supported yet.
+      self._fail_on_non_default_windowing(raw_data)
+      element_type = raw_data.element_type
+      column_type_mapping = {}
+      if (isinstance(element_type, RowTypeConstraint) or
+          native_type_compatibility.match_is_named_tuple(element_type)):
+        column_type_mapping = self._map_column_names_to_types(
+            row_type=element_type)
+        # convert Row or NamedTuple to Dict
+        raw_data = (
+            raw_data
+            | ConvertNamedTupleToDict().with_output_types(
+                Dict[str, typing.Union[tuple(column_type_mapping.values())]]))
+        # AnalyzeAndTransformDataset raise type hint since this is
+        # schema'd PCollection and the current output type would be a
+        # custom type(NamedTuple) or a beam.Row type.
+      else:
+        column_type_mapping = self._map_column_names_to_types_from_transforms()
+      raw_data_metadata = self.get_raw_data_metadata(
+          input_types=column_type_mapping)
+      # Write untransformed metadata to a file so that it can be re-used
+      # during Transform step.
+      metadata_io.write_metadata(
+          metadata=raw_data_metadata,
+          path=os.path.join(self.artifact_location, RAW_DATA_METADATA_DIR))
+    else:
+      # Read the metadata from the artifact_location.
+      if not os.path.exists(os.path.join(
+          self.artifact_location, RAW_DATA_METADATA_DIR, SCHEMA_FILE)):
+        raise FileNotFoundError(
+            "Raw data metadata not found at %s" %
+            os.path.join(self.artifact_location, RAW_DATA_METADATA_DIR))
+      raw_data_metadata = metadata_io.read_metadata(
+          os.path.join(self.artifact_location, RAW_DATA_METADATA_DIR))
+
+    # To maintain consistency by outputting numpy array all the time,
+    # whether a scalar value or list or np array is passed as input,
+    #  we will convert scalar values to list values and TFT will ouput
+    # numpy array all the time.
+    if not self.is_input_record_batches:
+      raw_data |= beam.ParDo(ConvertScalarValuesToListValues())
```

Comment:
We need to know before pipeline runtime whether it is a record batch or dict to determine the right schema.

So during pipeline construction, I use this flag to construct respective schema. I don't think this can be inferable later